neurom/features/morphology.py

```diff
@@ -177,26 +180,168 @@ def trunk_angles(morph, neurite_type=NeuriteType.all):
     if len(vectors) == 0:
         return []
 
-    def _sort_angle(p1, p2):
-        """Angle between p1-p2 to sort vectors."""
-        ang1 = np.arctan2(*p1[::-1])
-        ang2 = np.arctan2(*p2[::-1])
-        return ang1 - ang2
+    if sort_along:
+        # Sorting angles according to the given plane
+        sort_coords = str_to_plane(sort_along)
+        order = np.argsort(
+            np.array(
+                [
+                    morphmath.angle_between_projections(i / np.linalg.norm(i), [0, 1])
+                    for i in vectors[:, sort_coords]
+                ]
+            )
+        )
+
+        vectors = vectors[order]
+
+    # Select coordinates to consider
+    if coords_only:
+        coords = str_to_plane(coords_only)
+        vectors = vectors[:, coords]
+
+    # Compute angles between each trunk and the next ones
+    cycling_vectors = np.vstack([vectors, vectors])
+    angles = [
+        (num_i, [
+            morphmath.angle_between_vectors(i, j)
+            for j in cycling_vectors[num_i: num_i + len(vectors)]
+        ])
+        for num_i, i in enumerate(vectors)
+    ]
+
+    if consecutive_only:
+        angles = [i[1][-1] for i in angles if i[1]]
+    else:
+        angles = [i[1] for i in angles]
+
+    return angles
+
+
+@feature(shape=(...,))
+def trunk_angles_inter_types(
+    morph,
+    source_neurite_type=NeuriteType.apical_dendrite,
+    target_neurite_type=NeuriteType.basal_dendrite,
+    closest_component=None,
+):
+    """Calculate the angles between the trunks of the morph of a given type to another type.
+
+    For each couple of neurite, an array with 3 elements is returned:
+    * the absolute 3d angle between the two vectors.
+    * the elevation angle (or polar angle) between the two vectors.
+    * the azimuth angle between the two vectors.
+
+    If ``closest_component`` is not ``None``, only one element is returned for each neurite of
+    source type:
+    * if set to 0, the one with the lowest absolute 3d angle is returned.
+    * if set to 1, the one with the lowest absolute elevation angle is returned.
+    * if set to 2, the one with the lowest absolute azimuth angle is returned.
+    """
+    source_vectors = np.array(trunk_vectors(morph, neurite_type=source_neurite_type))
+    target_vectors = np.array(trunk_vectors(morph, neurite_type=target_neurite_type))
+
+    # In order to avoid the failure of the process in case the neurite_type does not exist
+    if len(source_vectors) == 0 or len(target_vectors) == 0:
+        return []
+
+    angles = np.array([
+        np.vstack([
+            np.concatenate(
+                [
+                    [morphmath.angle_between_vectors(i, j)],
+                    morphmath.spherical_from_vector(j) - morphmath.spherical_from_vector(i)
+                ]
+            )
+            for j in target_vectors
+        ])
+        for i in source_vectors
+    ])
```

Comment:
Please check that I got the order of the dimensions right:
```python
angles = np.empty((len(source_vectors), len(target_vectors), 3)), dtype=np.float32)

for i, source in enumerate(source_vectors):
    for j, target in enumerate(target_vectors):
        angles[i, j, 0] = morphmath.angle_between_vectors(source, target)]
        angles[i, j, 1] = morphmath.spherical_from_vector(target) - morphmath.spherical_from_vector(source)
```

Reply:
Isn't it slower because of the multiple assignments?

Reply:
I believe it's quite the opposite because you preallocate all the memory you are going to need.